src/main/java/org/commcare/formplayer/application/FormSubmissionController.java

```diff
@@ -0,0 +1,323 @@
+
+        // package additional args to pass to category timing helper
+        Map<String, String> extras = new HashMap();
+        extras.put(Constants.DOMAIN_TAG, submitRequestBean.getDomain());
+
+        return new FormSubmissionContext(request, submitRequestBean, formEntrySession, extras);
+    }
+
+    /**
+     * Execute a step in the process and return an error response to halt processing or
+     * and empty response to continue.
+     *
+     * @param request The HTTP Request object
+     * @param step A supplier object that performs one unit of form processing and returns
+     *             a SubmitResponseBean.
+     * @return Empty Optional if the processing should continue otherwise an Optional containing the
+     *          error response.
+     */
+    private Optional<SubmitResponseBean> executeStep(HttpServletRequest request, ProcessingStep step) {
+        SubmitResponseBean response = null;
+        try {
+            response = step.execute();
+        } catch (Exception e) {
+            response = getErrorResponse(
+                    request, "error",
+                    e.getMessage(), e);
+        }
+        if (response.getStatus().equals(Constants.SYNC_RESPONSE_STATUS_POSITIVE)) {
+            step.recordCheckpoint();
+            return Optional.empty();  // continue processing
+        }
+        log.debug(String.format("Aborting execution of processing steps after error in step: %s", step));
+        return Optional.of(response);
+    }
+
+    private SubmitResponseBean validateAnswers(FormSubmissionContext context) {
+        categoryTimingHelper.timed(
+                Constants.TimingCategories.VALIDATE_SUBMISSION,
+                () -> validateSubmitAnswers(context),
+                context.getMetricsTags()
+        );
+        if (!context.getResponse().getStatus().equals(Constants.SYNC_RESPONSE_STATUS_POSITIVE)
+                || !context.getRequest().isPrevalidated()) {
+            return context.error(Constants.ANSWER_RESPONSE_STATUS_NEGATIVE);
+        }
+        return context.success();
+    }
+
+    private SubmitResponseBean getErrorResponse(
+            HttpServletRequest request,
+            String status,
+            String message,
+            Throwable exception) {
+        SubmitResponseBean responseBean = new SubmitResponseBean(status);
+        NotificationMessage notification = new NotificationMessage(
+                message,
+                true,
+                NotificationMessage.Tag.submit);
+        responseBean.setNotification(notification);
+        logNotification(notification, request);
+        log.error(message, exception);
+        return responseBean;
+    }
+
+    private SubmitResponseBean processFormXml(FormSubmissionContext context) {
+        try {
+            restoreFactory.setAutoCommit(false);
+            processXmlInner(context);
+
+            String response = submitService.submitForm(
```

Comment:
I did originally have it that way but re-combined it to avoid changing functionality. I also wondered if it wasn't better to treat these as 'atomic' web apps does not have the same level of sophistication in terms of re-submitting forms in the background. If the user navigated away from the page their local db state would be different from HQ and there would be no way to submit the form to HQ.

Reply:
that resonsates with me. thanks!